src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -40,8 +108,8 @@ private void InitializeCancellationToken(TimeSpan timeout)
             }
             else if (timeout > TimeSpan.Zero)
             {
-                _cts = new CancellationTokenSource(timeout);
-                _cancellationToken = _cts.Token;
+                _cts = new CancellationTokenSource();
+                _cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(_cts.Token, TimeoutTokenSource.FromTimeout((int)timeout.TotalMilliseconds)).Token;
```

Comment:
Since you're not cleaning this up anyway, consider trying out:
```C#
_cts = new CancellationTokenSource();
_cancellationToken = _cts.Token;
TimeoutTokenSource.FromTimeout((int)timeout.TotalMilliseconds)).Token.Register(s => ((CancellationTokenSource)s).Cancel(), _cts);
```
and seeing whether that helps with performance.  It should have the same behavior but should save a couple of allocations.